### PR TITLE
Trust dns client cli

### PR DESCRIPTION
**Pull request by bluejekyll**
This is a new CLI for trust-dns-client that mimics the `dig` command:

```shell
$> cargo install --all-features --path util --bin dns
...
```

The options are:

```shell
$> dns -h
trust dns client 0.21.2
A CLI interface for the trust-dns-client

USAGE:
    dns [OPTIONS] --nameserver <NAMESERVER> <SUBCOMMAND>

OPTIONS:
    -a, --alpn <ALPN>
            For TLS, HTTPS, and QUIC a custom ALPN code can be supplied

        --class <CLASS>
            The Class of the record [default: IN]

        --debug
            Enable debug and all logging

        --do-not-verify-nameserver-cert
            DANGER: do not verify remote nameserver

        --error
            Enable error logging

    -h, --help
            Print help information

        --info
            Enable info + warning + error logging

    -n, --nameserver <NAMESERVER>
            Specify a nameserver to use, ip and port e.g. 8.8.8.8:53 or \[2001:4860:4860::8888\]:53
            (port required)

    -p, --protocol <PROTOCOL>
            Protocol type to use for the communication [default: udp] [possible values: udp, tcp,
            tls, https, quic]

    -t, --tls-dns-name <TLS_DNS_NAME>
            TLS endpoint name, i.e. the name in the certificate presented by the remote server

    -V, --version
            Print version information

        --warn
            Enable warning + error logging

    -z, --zone <ZONE>
            Zone, required for dynamic DNS updates, e.g. example.com if updating www.example.com

SUBCOMMANDS:
    append           Append record data to a record set
    create           Create a new record in the target zone
    delete-record    Delete a single record from a zone, the data must match the record
    help             Print this message or the help of the given subcommand(s)
    notify           Notify a nameserver that a record has been updated
    query            Query a name server for the record of the given type
```

This supports all the protocols supported by trust-dns, `udp`, `tcp`, `tls`, `https`, and `quic`. 

Here's an example query:

```shell
$> dns -p https -t cloudflare-dns.com -n 1.1.1.1:443 query www.example.com. SOA
; using https:1.1.1.1:443 dns_name:cloudflare-dns.com
; sending query: www.example.com. IN SOA
; received response
; header 0:RD,RA:NoError:QUERY:0/1/1
; edns version: 0 dnssec_ok: false max_payload: 1232 opts: 0
; query
;; name: www.example.com. type: SOA class: IN
; answers 0
; nameservers 1
example.com. 3600 IN SOA ns.icann.org. noc.dns.icann.org. 2022040401 7200 3600 1209600 3600
; additionals 1
```

Fixes: #1663

### CODE REVIEW

**Comment by bluejekyll** [comment]
@djc if you have some time just to look at the library changes, it would be greatly appreciated.

**Comment by bluejekyll** [comment]
ping @djc, I'd love to land this, it's very useful for testing...

File: crates/client/src/serialize/txt/parse_rdata.rs [hunk 1]
```diff
@@ -22,12 +22,40 @@ use crate::proto::rr::dnssec::rdata::DNSSECRData;
 use crate::rr::{Name, RData, RecordType};
 use crate::serialize::txt::rdata_parsers::*;
 
-pub(crate) trait RDataParser: Sized {
+use crate::serialize::txt::zone_lex::Lexer;
```
Comment (djc) [inline crates/client/src/serialize/txt/parse_rdata.rs:25]: Feels like this `use` and the one below should be grouped with the block above?

**Review by djc — COMMENTED** [review]
Sorry for failing to review this sooner. I generally don't have much time on weekends to do larger reviews, so couldn't respond to your query from yesterday sooner.

File: crates/client/src/serialize/txt/parse_rdata.rs (hunk 1 above)
Comment (bluejekyll) [reply crates/client/src/serialize/txt/parse_rdata.rs:25]: Hm. That might have been a mistake. No worries on the timeline. I’ll post a patch to follow this up. I just needed to get these things bound together for some other testing. It was getting unwieldy keeping all the changes straight.

**Pull request merged**